test/bft_tester.py

```diff
@@ -0,0 +1,252 @@
+# Concord
+#
+# Copyright (c) 2019 VMware, Inc. All Rights Reserved.
+#
+# This product is licensed to you under the Apache 2.0 license (the "License").
+# You may not use this product except in compliance with the Apache 2.0 License.
+#
+# This product may include a number of subcomponents with separate copyright
+# notices and license terms. Your use of these subcomponents is subject to the
+# terms and conditions of the subcomponent's license, as noted in the LICENSE
+# file.
+
+# Add the pyclient directory to $PYTHONPATH 
+
+import sys
+
+import os
+import copy
+import os.path
+import random
+import subprocess
+from collections import namedtuple
+import tempfile
+import trio
+
+sys.path.append(os.path.abspath("../util/pyclient"))
+
+import bft_config
+import bft_client
+import bft_metrics_client
+import bft_metrics
+
+from bft_tester_exceptions import AlreadyRunningError, AlreadyStoppedError 
+
+TestConfig = namedtuple('TestConfig', [
+    'n',
+    'f',
+    'c',
+    'num_clients',
+    'key_file_prefix',
+    'start_replica_cmd'
+])
+
+MAX_MSG_SIZE = 64*1024 # 64k
+REQ_TIMEOUT_MILLI = 5000
+RETRY_TIMEOUT_MILLI = 250
+METRICS_TIMEOUT_SEC = 5
+
+# TODO: This is not generic, but is required for use by SimpleKVBC. In the
+# future we will likely want to change how we determine the lengths of keys and
+# values, make them parameterizable, or generate keys in the protocols rather
+# than tester. For now, all keys and values must be 21 bytes.
+KV_LEN = 21
+
+class BftTester:
+    """A library to help testing sbft with a SimpleKVBC state machine"""
+
+    def __enter__(self):
+        """context manager method for 'with' statements"""
+        return self
+
+    def __exit__(self, *args):
+        """context manager method for 'with' statements"""
+        for client in self.clients.values():
+            client.__exit__()
+        self.metrics.__exit__()
+        for proc in self.procs.values():
+            proc.kill()
+            proc.wait()
+
+    def __init__(self, config):
+        self.origdir = os.getcwd()
+        self.config = config
+        self.testdir = tempfile.mkdtemp()
+        print("Running test in {}".format(self.testdir))
+        self.builddir = os.path.abspath("../build")
+        self.toolsdir = os.path.join(self.builddir, "tools")
+        self.procs = {}
+        self.replicas = [bft_config.Replica(i, "127.0.0.1", 3710 + 2*i) 
+                for i in range(0, self.config.n)]
+        self.alpha = [i for i in range(65, 91)]
+        self.alphanum = [i for i in range(48, 58)]
+        self.alphanum.extend(self.alpha)
+        self.keys = self._create_keys()
+        os.chdir(self.testdir)
+        self._generate_crypto_keys()
+        self.clients = {}
+        self.metrics = None
+
+    def _generate_crypto_keys(self):
```

Comment:
I would add TLS certificates creation here as well.

Reply:
We only support UDP transport at the moment for these tests. When we add support for TLS I'll add in the key generation.